thinc/layers/dropout.py

```diff
@@ -18,40 +17,43 @@ def Dropout(rate: float = 0.0) -> Model[InT, InT]:
     return Model("dropout", forward, attrs={"dropout_rate": rate, "is_enabled": True})
 
 
-# We're getting type hell here, I think because of the instance checks?
-# It's sort of painful, because I think this confused the types of other
-# layers that are trying to use dropout.
-# I've relaxed the types for now, but it'd be good to understand what's wrong
-# here.
-def forward(model: Model, X, is_train: bool) -> Tuple[Any, Callable]:
+def forward(
+    model: Model[InT, InT], X: InT, is_train: bool
+) -> Tuple[InT, Callable]:
     rate = model.attrs["dropout_rate"]
     is_enabled = model.attrs["is_enabled"] and is_train
     if rate == 0 or not is_enabled:
-        return X, lambda dY: dY
+        unchanged_return_value, backprop = X, lambda dY: dY
+        return_value = cast(InT, unchanged_return_value)
     elif isinstance(X, Ragged):
-        return _dropout_ragged(model, X, is_train)
+        ragged_return_value, backprop = _dropout_ragged(model, X, is_train)
+        return_value = cast(InT, ragged_return_value)
     elif isinstance(X, Padded):
-        return _dropout_padded(model, X, is_train)
-    elif isinstance(X, list):
-        return _dropout_lists(model, X, is_train)
+        padded_return_value, backprop = _dropout_padded(model, X, is_train)
+        return_value = cast(InT, padded_return_value)
+    elif isinstance(X, Sequence):
+        list_return_value, backprop = _dropout_lists(model, X, is_train)
+        return_value = cast(InT, list_return_value)
     else:
-        return _dropout_array(model, X, is_train)
+        array_return_value, backprop = _dropout_array(model, cast(ArrayXd, X), is_train)
+        return_value = cast(InT, array_return_value)
+    return return_value, backprop
 
 
 def _dropout_array(
-    model: Model[ArrayT, ArrayT], X: ArrayT, is_train: bool
-) -> Tuple[ArrayT, Callable]:
+    model: Model[InT, InT], X: ArrayXd, is_train: bool
+) -> Tuple[ArrayXd, Callable]:
```

Comment:
I get why we want/need to change the typing of `model`, but why are we not using `ArrayT` anymore for `X` and the for the output tuple?

Reply:
I judged it to be unnecessary complexity, because we are casting on the way into and out of the `_dropout_array()` method as it is and this method will never be called from outside the module. So there is nowhere where the variance that `ArrayT` gave could be used.

Reply:
gotcha!